lib/logstash/config/config_ast.rb

```diff
@@ -211,13 +216,7 @@ def compile
         return "start_input(#{variable_name})"
       when "filter"
         return <<-CODE
-          events = events.flat_map do |event|
-            next [] if event.cancelled?
-
-            new_events = []
-            #{variable_name}.filter(event){|new_event| new_events << new_event}
-            event.cancelled? ? new_events : new_events.unshift(event)
-          end
+          events = #{variable_name}.multi_filter(events)
         CODE
```

Comment:
for event cancellation, see bottom of https://github.com/elastic/logstash/pull/2869#issuecomment-85132637